### ISSUE TITLE
Clarification on epoch based 'at least once' pipeline for distributed, workers > 1, batched scenario

### ISSUE THREAD

**Issue by rwightman** [webdataset/webdataset#169]
I've been working with 0.2.x recently updating a project that was using 0.1 and adding support to another, I have a rough template for a custom pipeline but there are some details I am bit concerned about. I feel a pipeline similar to this would be valuable to add to documentation at some point (as an alternative to the resampling distributed example). 

This is an epoch based pipeline, with the goal of 'each sample seen at least once per-epoch'. The eval path in this example is single-node where the train is multi. The setup of the pipeline is below, the big questions are regarding the use of the pipeline using `with_epochs` to acheive 'at least once' 

``` python
    pipeline = [wds.SimpleShardList(input_shards)]
    # at this point we have an iterator over all the shards
    if is_train:
        pipeline.extend([
            wds.detshuffle(
                  bufsize=_SHARD_SHUFFLE_SIZE,
                  initial=_SHARD_SHUFFLE_INITIAL,
                  seed=seed),
            wds.split_by_node,
            wds.split_by_worker,
            # at this point, we have an iterator over the shards assigned to each worker at each node
            wds.tarfile_to_samples(handler=log_and_continue),
            wds.shuffle(
                bufsize=_SAMPLE_SHUFFLE_SIZE,
                initial=_SAMPLE_SHUFFLE_INITIAL,
                rng=random.Random(seed)),
            #wds.repeatedly,  # repeatedly here to avoid dropping samples at end of iteration for batching?
        ])
    else:
        pipeline.extend([
            wds.split_by_worker,
            # at this point, we have an iterator over the shards assigned to each worker
            wds.tarfile_to_samples(handler=log_and_continue),
        ])
    pipeline.extend([
        wds.select(filter_no_caption),
        wds.decode("pilrgb", handler=log_and_continue),
        wds.rename(image="jpg;png", text="txt"),
        wds.map_dict(image=preprocess_img, text=preprocess_txt),
        wds.to_tuple("image", "text"),
        wds.batched(batch_size, partial=not is_train),
    ])
```

For using with_epochs, I have two schemes that appear to work (but haven't confirmed by checking all samples yet)

In this first one, if we apply `with_epochs` on dataset (initial data pipeline) we need to calculate num batches per worker as each worker is iterating over it's copy of dataset correct?

``` python
    dataset = wds.DataPipeline(*pipeline)
    if is_train:
        global_batch_size = batch_size * world_size
        num_batches = math.ceil(num_samples / global_batch_size)
        num_workers = max(1, workers)
        num_worker_batches = math.ceil(num_batches / num_workers)  # per dataloader worker
        num_batches = num_worker_batches * num_workers
        num_samples = num_batches * global_batch_size
        dataset = dataset.with_epoch(num_worker_batches)  # each worker is iterating over this
    else:
        # last batches are partial, eval is done on single (master) node
        num_batches = math.ceil(num_samples / batch_size)

    dataloader = wds.WebLoader(dataset, batch_size=None, shuffle=False, num_workers=workers)
```

Alternatively it appears we could apply `with_epochs` on dataloader instead, is there an efficiency downside to this? I assume that I'd still have to round num_batches to take num_workers into consideration so that each worker will have a equivalent # of batches and we won't leave some batches on one of the workers?

``` python
    dataset = wds.DataPipeline(*pipeline)
    dataloader = wds.WebLoader(dataset, batch_size=None, shuffle=False, num_workers=workers)
    if is_train:
        global_batch_size = batch_size * world_size
        num_batches = math.ceil(num_samples / global_batch_size)
        num_workers = max(1, workers)
        num_batches = math.ceil(num_batches / num_workers) * num_workers
        num_samples = num_batches * global_batch_size
        dataloader = dataloader.with_epoch(num_batches)
    else:
        # last batches are partial, eval is done on single (master) node
        num_batches = math.ceil(num_samples / batch_size)
```

And then one last concern, when using `batched()` as I am here, it looks like as batched is iterating over source it will leave samples unconsumed at the end of iteration. In conjunction with `with_epochs` at the end of the pipeline, I gather this will leave those samples unseen when the wrap around is performed? Should I be inserting  `repeatedly()` somewhere before batched to prevent this? Either after or before the wds.shuffle in the is_train branch for example?

**Comment by floschne** [comment]
Hi, thanks for posting this question, @rwightman. Did you find a solution regarding the optimal "position" where to put `with_epochs` ?

**Comment by tmbdev** [comment]
So, in short, your analysis sounds right, including your use of "repeatedly" to try to fill up epochs. The recommended way of using WebDataset for large scale, distributed training is with shard resampling.

Since many people didn't want to have to convert their existing "epoch-based" and indexed Dataset code to resampled streaming code, there is now a separate API that provides an indexed Dataset interface to WebDataset. You get it by "import wids". You can see how it works in the ./examples subdirectory.
